From 0f1f01bf1799c0a407f2aa773ce1ad7ff0e277b9 Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Fri, 29 Aug 2014 19:01:18 -0400
Subject: [net] bridge: Introduce BR_PROMISC flag

Message-id: <1409338885-30414-7-git-send-email-vyasevic@redhat.com>
Patchwork-id: 89906
O-Subject: [RHEL7 PATCH 06/13] bridge: Introduce BR_PROMISC flag
Bugzilla: 896669
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>

From: Vlad Yasevich <vyasevic@redhat.com>

Introduce a BR_PROMISC per-port flag that will help us track if the
current port is supposed to be in promiscuous mode or not.  For now,
always start in promiscuous mode.

Acked-by: Michael S. Tsirkin <mst@redhat.com>
Signed-off-by: Vlad Yasevich <vyasevic@redhat.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit f3a6ddf15209cfad90b83e04168f42a5d9c8cc17)
Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/bridge/br_if.c b/net/bridge/br_if.c
index 5ccbe7e..4d6e247 100644
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -238,7 +238,7 @@ static struct net_bridge_port *new_nbp(struct net_bridge *br,
  p->path_cost = port_cost(dev);
  p->priority = 0x8000 >> BR_PORT_BITS;
  p->port_no = index;
- p->flags = BR_LEARNING | BR_FLOOD;
+ p->flags = BR_LEARNING | BR_FLOOD | BR_PROMISC;
  br_init_port(p);
  p->state = BR_STATE_DISABLED;
  br_stp_port_timer_init(p);
diff --git a/net/bridge/br_private.h b/net/bridge/br_private.h
index 7fb3048..0268d83 100644
--- a/net/bridge/br_private.h
+++ b/net/bridge/br_private.h
@@ -174,6 +174,7 @@ struct net_bridge_port
 #define BR_LEARNING  0x00000020
 #define BR_FLOOD  0x00000040
 #define BR_AUTO_MASK (BR_FLOOD | BR_LEARNING)
+#define BR_PROMISC  0x00000080
 
 #ifdef CONFIG_BRIDGE_IGMP_SNOOPING
  struct bridge_mcast_query ip4_query;
@@ -199,6 +200,7 @@ struct net_bridge_port
 };
 
 #define br_auto_port(p) ((p)->flags & BR_AUTO_MASK)
+#define br_promisc_port(p) ((p)->flags & BR_PROMISC)
 
 #define br_port_exists(dev) (dev->priv_flags & IFF_BRIDGE_PORT)
 
-- 
1.7.1